### PR TITLE
ref(sentry-metrics): Add MetricsKeyIndexer table 

### PR DESCRIPTION
**Pull request by MeredithAnya**
**context**
This [PR](https://github.com/getsentry/sentry/pull/28431) added the metrics consumer (take messages from relay, translate it, and pass it along to snuba) but just used redis as a dummy indexer to store the string -> int (and vice versa) translations. The next phase is to add persistence, which will be Postgres.

**MetricsKeyIndexer**
This table will ultimately be on its own postgres instance. My understanding is that for the sake of development I can run this migration as normal. If that's not the case, well.. I'll need to figure out what I should be doing. 

**Schema**
From my understanding, the basic columns for this table are:
```python
string = str                 // the metric name, tag key, or tag value
id = int64             // the int version of the metric name, tag key or tag value

pk --> (id) /
btree index --> (string)
```
**BTree index on `string`**
We are going to want to be able to look up the `string` from the `id` as well as the `id` from the `string`, so it needs to be indexed both ways

**Sequence vs UUID**
We want the ids to be 64bit integers, (uuids would be twice as long at 128bit) so that ClickHouse is more efficient whenn querying. Additionally, the UUID type does not work with the bloom filter index[^1]. Ultimately we are going to want to pre-allocate multiple values at a time in redis so that we don't have to hit postgres every time we record a new metric (or tag key or value). For these reasons we can use a sequence (for now, could change later).

[^1]: needs to be verified, but heard from a reputable source: @fpacifici  

<hr/>

**[EDITED]:** Keeping below as historical context, but is outdated

It seems to me there are three main ways to use a sequence:
1. created independently from a table, literally just `CREATE SEQUENCE sequence_name`
2. created independently from a table, but owned by a column in the table `CREATE SEQUENCE sequence_name OWNED BY table.column_name` 
3. created behind the scenes when using an auto increment field, (what all our `id` fields do with `BoundedBigAutoField`)
   * Django doesn't seem to let you use an auto increment field that _isn't_ a primary key, even if you specify another field as the primary key. We could probably make a new Field to override this from happening but I wasn't sure this was worth it. 
   * If we used an auto increment field and tried to INSERT a different value (that wasn't generated from postgres) the auto increment isn't aware of that . **ex**. you are at value `5`, INSERT value = `10`, and then go back to using the increment, you'd be at `6` (and then at `10` youd get an `IntegrityError`)
   * I don't think there is a way to change the `INCREMENT` value from `1` so if we wanted it to use an `INCREMENT` of `100`, I dk how'd we do that. (maybe we'd just have to `ALTER SEQUENCE`)

For reason listed above I ~went~ started with option 2, but there are some caveats, issues:
* Unlike the auto increment, you must already have the `value` when you save the record. Since this is still in dev (and we aren't pre-allocating values yet), so I've kind of implemented that myself in the `save()` method for now. 
* For some reason, the test db doesn't seem to have the sequence created even though the `MetricsIndexerTable` is in there. I haven't figured out why, but in the mean time I have to create the sequence in `setUp` (and then drop in `tearDown`) for the tests. I don't have this problem just locally accessing postgres so idk.

### CODE REVIEW

File: src/sentry/conf/server.py [hunk 1]
```diff
@@ -1357,7 +1358,7 @@ def create_partitioned_queues(name):
 SENTRY_METRICS_SKIP_INTERNAL_PREFIXES = []  # Order this by most frequent prefixes.
 
 # Metrics product
-SENTRY_METRICS_INDEXER = "sentry.sentry_metrics.indexer.mock.MockIndexer"
+SENTRY_METRICS_INDEXER = "sentry.sentry_metrics.indexer.postgres.PGStringIndexer"
```
Comment (MeredithAnya) [inline src/sentry/conf/server.py:1361]: I don't really like the naming, I just wanted to make it different that the redis indexer, and I wasn't sure if we were ready to use this as the _actual_ string indexer yet

File: src/sentry/new_migrations/monkey/executor.py [hunk 2]
```diff
@@ -37,7 +37,9 @@ def _check_db_routing(migration):
         def _check_operations(operations):
             failed_ops = []
             for operation in operations:
-                if isinstance(operation, (FieldOperation, ModelOperation, RenameContentType)):
+                if isinstance(
+                    operation, (FieldOperation, ModelOperation, RenameContentType, IndexOperation)
```
Comment (MeredithAnya) [inline src/sentry/new_migrations/monkey/executor.py:41]: I had to add the `IndexOperation` here because otherwise I got the missing "`hints={'tables':..}` argument" for `AddIndex`. It seemed to me that since the `AddIndex` operation is model specific that I could put this here cc @wedamija

File: src/sentry/sentry_metrics/indexer/models.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    class Meta:
+        db_table = "sentry_metricskeyindexer"
+        app_label = "sentry"
+        indexes = [
+            models.Index(fields=["organization_id", "key"]),
+            models.Index(fields=["organization_id", "value"]),
+        ]
+
+    @classmethod
+    def get_next_values(cls, num: int):
```
Comment (MeredithAnya) [inline src/sentry/sentry_metrics/indexer/models.py:23]: This is not used right now but was my thought process for later, to test out get n number of next values. This does load all the integers which may not be ideal, so the other option would be to have the increment be larger and we just get the range from the last val to the next val e.g currval is 100 and next is 200, when we have 101 -200 or whatever

**Review by jjbayer — COMMENTED** [review]
This is way more advanced db design than anything I've ever done, but I wonder if we could meet the same requirements with the following model:

```python
class MetricsStringIndex(sentry.db.models.Model):
    organization_id = BoundedBigIntegerField()
    string = models.CharField(max_length=200)
    class Meta:
        unique_together = (("organization_id", "string"),)
```

The `id` field inherited from `Model` could serve as primary key _and_ integer value. This would provide a sequence and efficient `int -> string` lookups out of the box. The `unique_together` prevents duplicates and adds an index under the hood, enabling efficient `(org_id, string) -> int` lookups.

You would still be able to do 

```sql
SELECT nextval('sentry_metricsstringindex_id_seq') FROM generate_series(1, 100);
```

for pre-allocated values.

Maybe I'm completely missing some constraints though.

**Comment by MeredithAnya** [comment]
> This is way more advanced db design than anything I've ever done, but I wonder if we could meet the same requirements with the following model:
> 
> ```python
> class MetricsStringIndex(sentry.db.models.Model):
>     organization_id = BoundedBigIntegerField()
>     string = models.CharField(max_length=200)
>     class Meta:
>         unique_together = (("organization_id", "string"),)
> ```
> 
> The `id` field inherited from `Model` could serve as primary key _and_ integer value. This would provide a sequence and efficient `int -> string` lookups out of the box. The `unique_together` prevents duplicates and adds an index under the hood, enabling efficient `(org_id, string) -> int` lookups.
> 
> You would still be able to do
> 
> ```sql
> SELECT nextval('sentry_metricsstringindex_id_seq') FROM generate_series(1, 100);
> ```
> 
> for pre-allocated values.
> 
> Maybe I'm completely missing some constraints though.

If we want to make the `value` or `id` (whatever we wanna call it) the primary key then what you've outlined makes sense to me. Maybe the best thing for now is to use an auto increment field.

File: src/sentry/sentry_metrics/indexer/models.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        )
+        return connection.fetchall()
+
+    def save(self, *args: Any, **kwargs: Any) -> None:
+        if not self.value:
+            using = router.db_for_write(self.__class__)
+            connection = connections[using].cursor()
+            connection.execute("SELECT nextval('metricskeyindexer_value')")
+            next_val = connection.fetchone()[0]
+            self.value = next_val
```
[the remark this answers is not in the document]
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/indexer/models.py:40]: yeah I think that would be the case, which would be problematic. need the unique constraint on the org_id and string

File: src/sentry/new_migrations/monkey/executor.py (hunk 2 above)
Comment (wedamija) [reply src/sentry/new_migrations/monkey/executor.py:41]: That looks good to me

File: src/sentry/sentry_metrics/indexer/models.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    @classmethod
+    def get_next_values(cls, num: int) -> Any:
+        using = router.db_for_write(cls)
+        connection = connections[using].cursor()
+
+        connection.execute(
+            "SELECT nextval('sentry_metricskeyindexer_id_seq') from generate_series(1,%s)", [num]
+        )
+        return connection.fetchall()
```
Comment (wedamija) [inline src/sentry/sentry_metrics/indexer/models.py:30]: I don't have a lot of context on this project, how will we use the values from this sequence?

It vaguely looks like you want to reserve a range of ids, and then use those ids later on to create new rows. Is that the general idea?

Comment (MeredithAnya) [reply src/sentry/sentry_metrics/indexer/models.py:30]: @wedamija Sorry for not giving enough context in the PR description, I can go back an update in a bit but yeah:

> It vaguely looks like you want to reserve a range of ids, and then use those ids later on to create new rows. Is that the general idea?

Thats basically it. Eventually we want to have postgres be off the critical path, but in order to do that we need to know the ids ahead of time. What I am unsure about is what kind of ranges we are talking, is it 100, 1000, 10000? Since this metrics indexer will be used by metrics names, tag keys, and tag values, it could be a lot of writes for high cardinality tags

File: src/sentry/migrations/0235_add_metrics_indexer_table.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            ],
+            options={
+                "db_table": "sentry_metricskeyindexer",
+            },
+        ),
+        migrations.AddConstraint(
+            model_name="metricskeyindexer",
+            constraint=models.UniqueConstraint(
+                fields=("organization_id", "string"), name="unique_org_string"
+            ),
```
Comment (fpacifici) [inline src/sentry/migrations/0235_add_metrics_indexer_table.py:54]: Does this also add the index ?

File: src/sentry/sentry_metrics/indexer/models.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+from sentry.db.models import Model
+from sentry.db.models.fields.bounded import BoundedBigIntegerField
+
+
+class MetricsKeyIndexer(Model):  # type: ignore
+    __include_in_export__ = False
+
+    organization_id = BoundedBigIntegerField()
+    string = models.CharField(max_length=200)
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/models.py:13]: I am not sure I get how does the int autoincrementing field get created.

File: src/sentry/sentry_metrics/indexer/postgres.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    Provides integer IDs for metric names, tag keys and tag values
+    and the corresponding reverse lookup.
+    """
+
+    __all__ = ("record", "resolve", "reverse_resolve", "bulk_record")
+
+    def _bulk_record(self, org_id: int, unmapped_strings: Set[str]) -> List[Record]:
+        records = []
+        for string in unmapped_strings:
+            obj = MetricsKeyIndexer.objects.create(organization_id=org_id, string=string)
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/postgres.py:25]: Do we have a way to create objects in batch? It should be more efficient for postgres.

File: src/sentry/migrations/0235_add_metrics_indexer_table.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name="MetricsKeyIndexer",
+            fields=[
+                (
+                    "id",
+                    sentry.db.models.fields.bounded.BoundedBigAutoField(
+                        primary_key=True, serialize=False
```
Comment (fpacifici) [inline src/sentry/migrations/0235_add_metrics_indexer_table.py:40]: Isn't this supposed to be auto-increment ?

**Review by fpacifici — COMMENTED** [review]
Mostly some clarification on how the model is built. Otherwise seems good to me.

File: src/sentry/migrations/0235_add_metrics_indexer_table.py (hunk 6 above)
Comment (MeredithAnya) [reply src/sentry/migrations/0235_add_metrics_indexer_table.py:54]: yeah, and also the auto increment happens to the `id` because the default is set to the `nextval()` for the sequence

```
sentry=# \d sentry_metricskeyindexer
                                     Table "public.sentry_metricskeyindexer"
     Column      |          Type          |                               Modifiers
-----------------+------------------------+-----------------------------------------------------------------------
 id              | bigint                 | not null default nextval('sentry_metricskeyindexer_id_seq'::regclass)
 organization_id | bigint                 | not null
 string          | character varying(200) | not null
Indexes:
    "sentry_metricskeyindexer_pkey" PRIMARY KEY, btree (id)
    "unique_org_string" UNIQUE CONSTRAINT, btree (organization_id, string)
```

File: src/sentry/sentry_metrics/indexer/models.py (hunk 7 above)
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/indexer/models.py:13]: it's because I'm inheriting from the `Model` class that sets the id to the `BigBoundedAutoField` (which uses the pseudo type bigserial and thats what tells postgres to create the sequence) https://github.com/getsentry/sentry/blob/197cbb0375c9259b422c85704578763a40384a92/src/sentry/db/models/base.py#L85-L86

File: src/sentry/sentry_metrics/indexer/postgres.py (hunk 8 above)
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/indexer/postgres.py:25]: I wasn't using `bulk_create` before because it doesn't call `save()` (which I was using to set the value) but now that I've gone back to using an auto increment field I think we can use that.

**Comment by wedamija** [comment]
@MeredithAnya I'm curious about whether we need to use this sequence at all. What are the requirements for these ids? I'm wondering whether we could use UUIDs here

**Comment by MeredithAnya** [comment]
> @MeredithAnya I'm curious about whether we need to use this sequence at all. What are the requirements for these ids? I'm wondering whether we could use UUIDs here

@wedamija the ids have to be 64 bit integers and according to the maths (not done by me tho) we can't use a hashing system because the probability of collisions is too high for comfort

**Comment by wedamija** [comment]
> > @MeredithAnya I'm curious about whether we need to use this sequence at all. What are the requirements for these ids? I'm wondering whether we could use UUIDs here
> 
> @wedamija the ids have to be 64 bit integers and according to the maths (not done by me tho) we can't use a hashing system because the probability of collisions is too high for comfort

If we were hashing into 64 bit integers then yeah, I'd agree that the collision chance is too high. UUIDs are the usual solution for this, but if we only have 64 bits to play with then there's not much choice. Is there a doc I can read about this in general?

File: src/sentry/sentry_metrics/indexer/models.py (hunk 5 above)
Comment (wedamija) [reply src/sentry/sentry_metrics/indexer/models.py:30]: Looks good for now. Once we know how many ids we're allocation per second we can decide whether we need to do something more complex here. 

I'm not sure if there's a performance hit to calling `nextval` 10k times, as opposed to doing something like https://www.depesz.com/2008/03/20/getting-multiple-values-from-sequences/. Something we can possibly benchmark in the future.

File: src/sentry/sentry_metrics/indexer/postgres.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    and the corresponding reverse lookup.
+    """
+
+    __all__ = ("record", "resolve", "reverse_resolve", "bulk_record")
+
+    def _bulk_record(self, org_id: int, unmapped_strings: Set[str]) -> Any:
+        records = [
+            MetricsKeyIndexer(organization_id=org_id, string=string) for string in unmapped_strings
+        ]
+        return MetricsKeyIndexer.objects.bulk_create(records)
```
Comment (wedamija) [inline src/sentry/sentry_metrics/indexer/postgres.py:26]: If we hit unique constraint conflicts here what do we want to do? I assume you want this to be like a bulk `get_or_create`. I wrote something like this recently in `bulk_get_rule_status`.

tl;dr is that you can bulk create them with `ignore_conflicts=True`, but then you won't have the ids filled in. So you will then need to make a second query to fetch all the rows you just created.

File: src/sentry/sentry_metrics/indexer/postgres.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    """
+
+    __all__ = ("record", "resolve", "reverse_resolve", "bulk_record")
+
+    def _bulk_record(self, org_id: int, unmapped_strings: Set[str]) -> List[Record]:
+        records = []
+        for string in unmapped_strings:
+            obj = MetricsKeyIndexer.objects.create(organization_id=org_id, string=string)
+            records.append(obj)
+        return records
```
[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/sentry_metrics/indexer/postgres.py]: Oops, I thought I submitted this earlier and it was pending, ignore

File: src/sentry/sentry_metrics/indexer/models.py (hunk 7 above)
Comment (wedamija) [reply src/sentry/sentry_metrics/indexer/models.py:13]: Is it valuable to have a `date_added` column here? Not sure how space conscious we need to be, but it's often helpful

Comment (fpacifici) [reply src/sentry/sentry_metrics/indexer/models.py:13]: That's a good idea. The metrics team is going to build the part of the data model that manages TTL and expiration, so they should have that and the last read time as well. But they are still working on that, I don't think that part of the data model exists yet. Still having a date added can be useful.

File: src/sentry/sentry_metrics/indexer/models.py (hunk 5 above)
Comment (fpacifici) [reply src/sentry/sentry_metrics/indexer/models.py:30]: I bet it will be desirable to avoid calling nextval 10k times (that would be 10k writes I believe). We may not strictly need a sequence at that point but just a counter.
I think we could discuss the solutions then. At this point I am not sure it is very useful to have this method at this stage.  Probably better removing it for now in case somebody decided to start depending on it for some reason.

File: src/sentry/sentry_metrics/indexer/postgres.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+    def bulk_record(self, org_id: int, strings: List[str]) -> Dict[str, int]:
+        # first look up to see if we have any of the values
+        records = MetricsKeyIndexer.objects.filter(organization_id=org_id, string__in=strings)
+        result = defaultdict(int)
+
+        for record in records:
+            result[record.string] = record.id
+
+        # find the unmapped strings
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/postgres.py:36]: you probably do not need this comment. What it does is pretty clear.

File: src/sentry/sentry_metrics/indexer/postgres.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        for new in new_mapped:
+            result[new.string] = new.id
+
+        return result
+
+    def record(self, org_id: int, string: str) -> int:
+        """Store a string and return the integer ID generated for it
+
+        With every call to this method, the lifetime of the entry will be
+        prolonged.
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply src/sentry/sentry_metrics/indexer/postgres.py:49]: I don't think this comment is relevant in the postgres implementation. It would have been in redis.
You can remove it.

File: tests/sentry/snuba/test_metrics.py [hunk 14]
```diff
@@ -1,292 +1,341 @@
...
-            "field": [
-                "sum(session)",
-                "count_unique(user)",
-                "p95(session.duration)",
-            ],
-        }
-    )
-    query_definition = QueryDefinition(query_params)
-    snuba_queries = SnubaQueryBuilder(PseudoProject(1, 1), query_definition).get_snuba_queries()
+class MetricsSnubaTest(TestCase):
```
Comment (fpacifici) [inline tests/sentry/snuba/test_metrics.py:38]: Is this test meant to be an end to end test for the metrics product ?
It seems to me the old version was a unit test for the abstractions the built over query and query result. If that is the case, it does not seem it needs to depend on postgres to work.
Unit tests on small modules have a big value as they would tell you immediately what is broken if a regression appears without the need to bring in a lot of dependencies.  So if this was meant to be like that, I would not add the persistence aspect to it.

Comment (MeredithAnya) [reply tests/sentry/snuba/test_metrics.py:38]: @jjbayer you may be the best person to answer this question. I can leave the mock indexer for this test in place

Comment (fpacifici) [reply tests/sentry/snuba/test_metrics.py:38]: Please leave the mock in place to start with. Then we can sync with @jjbayer about this

**Review by fpacifici — APPROVED** [review]
Looks good to me. Please see the comment about the mock

File: tests/sentry/snuba/test_metrics.py (hunk 14 above)
Comment (jjbayer) [reply tests/sentry/snuba/test_metrics.py:38]: > It seems to me the old version was a unit test for the abstractions the built over query and query result.
Yes, this was the intention, although I never enforced the mock (the test simply assumed that the `indexer` service was configured to the mock indexer).
